File: elixir/lib/2018/05_alchemical_reduction/aoc201805.ex
defmodule AOC2018.Day05 do
  @moduledoc """
  Advent of Code 2018, day 5: Alchemical Reduction
  """
  require AOC

  @doc """
  Parse input
  """
  def parse(puzzle_input), do: puzzle_input |> String.to_charlist()

  @doc """
  Solve part 1
  """
  def part1(input), do: input |> react() |> length()

  @doc """
  Solve part 2, with concurrency
  """
  def part2(input) do
    ?a..?z
    |> Task.async_stream(fn type -> input |> remove(type) |> react() |> length() end)
    |> Stream.map(fn {:ok, result} -> result end)
    |> Enum.min()
  end

  @doc """
  React a polymer

  ## Examples:

      iex> react('aBbA')
      ''

      iex> react('aabBCC')
      'aaCC'

      iex> react('aAbbCCDd')
      'bbCC'
  """
  def react(polymer), do: react(polymer, [])
  def react([head | rest], []), do: react(rest, [head])
  def react([], acc), do: acc |> Enum.reverse()
  def react([head | rest], [prev | acc]) when abs(head - prev) == 32, do: react(rest, acc)
  def react([head | rest], acc), do: react(rest, [head | acc])

  @doc """
  Remove all units of the given type from a polymer

  ## Examples:

      iex> remove('abBA', ?a)
      'bB'

      iex> remove('abBA', ?B)
      'aA'

      iex> remove('abBA', ?c)
      'abBA'

      iex> remove('aaAa', ?a)
      ''
  """
  def remove(polymer, type), do: Enum.reject(polymer, fn char -> rem(char - type, 32) == 0 end)

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/2018/05_alchemical_reduction/aoc201805.py
"""AoC 5, 2018: Alchemical Reduction."""

# Standard library imports
import pathlib
import sys

# Third party imports
import numpy as np


def parse_data(puzzle_input):
    """Parse input."""
    polymer = np.array(
        [
            ord(letter) - 96 if letter >= "a" else 64 - ord(letter)
            for letter in puzzle_input
        ]
    )
    return reduce(polymer)


def part1(polymer):
    """Solve part 1."""
    return len(polymer)


def part2(polymer):
    """Solve part 2."""
    elements = {abs(unit) for unit in polymer}
    polymer_lengths = [
        len(reduce(polymer[np.abs(polymer) != element])) for element in elements
    ]
    return min(polymer_lengths)


def reduce(polymer):
    """Reduce one polymer as much as possible"""
    while True:
        reductions = polymer[:-1] + polymer[1:]
        if not np.any(reductions == 0):
            return polymer

        idx, *_ = np.where(reductions == 0)
        reduce_idx = idx[np.concatenate(([True], np.diff(idx) != 1))]
        polymer[reduce_idx] = 99
        polymer[reduce_idx + 1] = 99
        polymer = polymer[polymer != 99]


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))
